optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java

```diff
@@ -102,13 +103,25 @@ public String getName() {
      */
     public static void defineClonerFor(ClassCreator classCreator, SolutionDescriptor<?> solutionDescriptor,
             List<Class<?>> solutionClassList,
-            Map<Class<?>, GizmoSolutionOrEntityDescriptor> memoizedSolutionOrEntityDescriptorMap) {
+            Map<Class<?>, GizmoSolutionOrEntityDescriptor> memoizedSolutionOrEntityDescriptorMap,
+            Set<Class<?>> deepClonedClassSet) {
+        List<Class<?>> deepCloneClassesThatAreNotSolutionList =
+                deepClonedClassSet.stream()
+                        .filter(clazz -> !solutionClassList.contains(clazz))
+                        .filter(clazz -> !clazz.isInterface() && !Modifier.isAbstract(clazz.getModifiers()))
+                        .sorted(instanceOfComparator)
+                        .distinct()
+                        .collect(Collectors.toList());
```

Comment:
If you put it into a `Set`, you won't need the `distinct()`.
But iterating over the Set will be slower than interating over the List would have been.
Was that the intention here?

Reply:
If the intention is to have order, you can use `Collectors.toCollection(...)` and provide a `TreeSet`.

Reply:
TBH, the comparator fails transitively: "Animal > Robot, Robot > Zebra", but "Zebra > Animal" (unrelated are sorted by name, related by isInstanceOf). We will need to use something like https://github.com/kiegroup/optaweb-employee-rostering/blob/master/optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/util/HierarchyTree.java  in order to decide what subclass to create properly. But I rather not create a new data structure for this; we might be able to calculate the "Root" classes (classes which are not subclasses of any other classes in the collection), construct a Map<Class<?>, List<Class<?>> that map each class in the collection to a list of it superclasses in order to define a better comparator).

As for why it needed:
We want:
```
if (x instanceof subclass1) {
...
} else if ((x instanceof subclass2) {
...
}  else if (x instanceof superclass) {
...
}
```
and not
```
if (x instanceof superclass) {
...
} else if ((x instanceof subclass1) {
...
}  else if (x instanceof subclass2) {
...
}
```

Reply:
Fixed;
for something like this:
```
    private interface Animal {
    }

    private interface Robot {
    }

    private interface Zebra extends Animal {
    }

    private interface RobotZebra extends Zebra, Robot {
    }
```
the comparator will sort it into
`[RobotZebra (2), Zebra (1), Robot (0), Animal (0)]`

(The number in parenthesis is the subclass level; root classes (which are not subclasses of anything) have a subclass level of 0; all other classes have a subclass level of the max subclass level of it parent classes + 1).